### ISSUE TITLE
XML-RPC introspection API support

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11547]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#80|
|Type|enhancement|
|Created|2003-07-15 01:01:50Z|

Attachments:

* [xmlrpc.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b88/b888b29826bb53dc531437e723738383d8339b56/688714645378a27a96ffb76213eb38d9141c5301.diff) (3621 bytes) - added by davep on 2003-08-13 14:37:43Z - 
* [xmlrpc.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b88/b888b29826bb53dc531437e723738383d8339b56/c067468173fc3c529138dcab2628f6ed27c2de74.diff) (6603 bytes) - added by davep on 2003-08-15 15:21:32Z - 
* [xmlrpc.3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b88/b888b29826bb53dc531437e723738383d8339b56/7d7e2acf84fc3dd9e50a965cbd9a71898ff897b3.diff) (7019 bytes) - added by davep on 2003-08-22 05:30:28Z - 
* [xmlrpc.4.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b88/b888b29826bb53dc531437e723738383d8339b56/15c230f437bd739cf7e136a4fc4e15a36b25d9fa.diff) (10504 bytes) - added by davep on 2003-08-25 13:26:19Z - 
* [xmlrpc.5.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b88/b888b29826bb53dc531437e723738383d8339b56/7b59784cb689b0744805649edbe8422baab423f9.diff) (13379 bytes) - added by davep on 2003-08-26 11:35:03Z - 
* [xmlrpc.6.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b88/b888b29826bb53dc531437e723738383d8339b56/13ebe2c0222fdc64d98113df882fb433f7b2798a.diff) (18142 bytes) - added by davep on 2003-08-27 04:48:24Z - 
* [xmlrpc.7.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b88/b888b29826bb53dc531437e723738383d8339b56/f86adf96da89e4acb7c50b23d6ae2ccfcea3385e.diff) (18316 bytes) - added by davep on 2003-08-27 07:20:50Z - 
* [xmlrpc.8.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b88/b888b29826bb53dc531437e723738383d8339b56/79792573df9705bdca84af07f15bd87ada538dce.diff) (18583 bytes) - added by davep on 2003-08-27 12:26:44Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__80 80
type__enhancement enhancement
reporter__itamarst itamarst
priority__low low
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__ 
time__1058230910000000 1058230910000000
changetime__1465005279957657 1465005279957657
version__ 
owner__ 
cc__spiv cc__davep cc__itamarst
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
XML-RPC has an informal introspection API
(http://ldp.kernelnotes.de/HOWTO/XML-RPC-HOWTO/xmlrpc-howto-interfaces.html)
which twisted.web.xmlrpc should support as an option.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
Here's the start of a patch to add this API.
I have implemented the system.listMethods function.

The patch adds a mix-in (XMLRPCIntrospection) that
will implement the introspection api.

The informal spec uses method names with periods.
To get around that issue, periods are converted to
double underscores for python instance method names.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
It occurs to me that maybe using the double-underscore
trick isn't so nice. Perhaps it would be better to add
support for adding named sub-handlers to an XMLRPC server
and add the introspection api under 'system'.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Sounds reasonable. And you'd register a subsystem for
"system.", not "system" (since xmlrpc doesn't require "." as
separator, other people may use other separators).

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
The subsystem/handler idea sounds good to me -- in the past
I've overridden _getFunction to replace dot with
double-underscore too.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
Here's a new version that uses the subsystem idea.
You use putChild to add a subhandler to an existing
server. I made the separator a class variable to it
can be overridden.

This patch also implements the system.methodHelp API.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
I'm not sure if putChild is a good method for this, since
it's intended for resources and XML-RPC does *not* map to
URLs, everything is same URL.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
davep is my new best friend

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
Here's a version that doesn't use putChild, but instead
some new API for XMLRPC. This leaves system.methodSignature
as the remaining API to be implemented. I'm just going to do
a simple 'signature registry' in the XMLRPCIntrospection 
object, unless someone thinks there is a better way.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
reflect.getsignature plus magic for classes etc

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
So are you proposing to create a new function in the
reflect module? I don't see this defined currently.

Note that the informal standard expects to get back
information about the types of both the return value
and the arguments:

  http://xmlrpc.usefulinc.com/doc/sysmethodsig.html

import xmlrpclib
server =
xmlrpclib.Server("http://xmlrpc-c.sourceforge.net/api/sample.php")
print server.system.methodHelp('system.methodSignature')
print server.system.methodSignature('system.methodSignature')

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
sorry, meant inspect module

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
Ok, so something like inspect.getargspec. But that just
returns argument names and default values. If we want to
implement the standard, we have to return the xml-rpc types
that can be passed/returned, and there's no way a python
inspection module is going to have that info.

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
I wonder how hard it would be to extract "@type str" and
similar from epytext docstrings...?

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
That seems like a reasonable route to take. You won't
be able to specify multiple signatures for a single
function, as the standard allows, but you avoid the
need for a central registry.

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
You could do multiple signatures for e.g.:

    def xmlrpc_foo(self, bar, baz, qux=99):
        """..."""

The 'foo' function can take arguments of ('bar', 'baz') or
('bar', 'baz', 'qux').  I won't mind if you don't bother
implementing this, though :)

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
Here's a version that implements system.methodSignature
using the @type idea. It doesn't currently handle varargs
or varkws, which aren't covered by the standard.

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Looks good to me!

(I wonder if it'd be better to use epydoc as a library to
extract information from docstrings, but that's probably
unnecessary complexity, so I'm happy).

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Looks pretty good... but! The @ syntax conflicts with
epydoc's usage (the types are different, for one thing.) How
about function attributes?

def xmlrpc_foo(self, bar): return 2 * bar
xmlrpc_foo.signature = [("bar", "integer")], "integer"
xmlrpc_foo.help = "multiply integer by two"

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
Ah, that could work, and would support multiple
signatures per function, too.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
Here's a new patch using method attributes. It also has
a doc patch to document xmlrpc sub-handlers in the xmlrpc
howto.

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
I think I'd rather see the epytext used rather than function
attributes, at least by default...

If the problem with "@type foo: int" is that xml-rpc expects
'integer' as a type, not 'int', then surely the introspector
could do the necessary mapping from python type names to
xml-rpc type names?  I'm probably missing something, but I
don't see any fatal flaws with the epytext approach...

I'm happy with whatever davep wants to implement, though,
seeing as he's doing all the hard work :)

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Actually, function attributes is probably a better approach
-- if someone (like me :) wants to use docstrings for that
data, they could just do:
    def xmlrpc_foo(self, bar):
        """...."""

then later do:
    initXMLRPCInfoFromDocstring(MyResource.xmlrpc_foo)

or something along those lines.

So, I'm happy with the function attributes patch.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
My patch uses __doc__ strings if no help attribute is set.
It seemed silly not to use it if it was available. Some
methods won't need different XML-RPC help strings.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
By the way, spiv, you posted message 666 in this issue.
I don't know if you want to take any precautions :)

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
I don't like epytext because it has overloaded meaning. On
the one hand it is docs for programmers reading the code, on
the other hand it is used for remote API. I hated the fact
Zope had code which behaved differently depending on if the
function had a docstring or not.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
Here's a patch I think is ready to go, with 
complete documentation.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
looks good. feel free to commit.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
hmm, one caveat. I'd be happier if instead of the way you do
it now, you'd do:

def addIntrospection(xmlrpc)::
    xmlrpc.putSunbhandler("system", XMLRPCIntrospection(xmlrpc))

on the top-level of the module and that's how people would
enable it.

Also, what happens if separator is *not* "."? OTOH, this is
probably unlikely, and people can still do it with a bit of
hackery if we change the API to look as above.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
Yeah, that way is better.

If the separator is not '.' then accessing
the introspection API will require using whatever
separator you have specified, 'system/listMethods'
or whatever.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
Ok, this should do it, then.

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
itamar: Regarding epytext, I generally think that docstrings
and comments should be used as follows:

    def func(args...):
        """Docstring: details on what this func is for and 
        how to call it.
        """
        # Comments about implementation details go here, 
        # rather than in docstrings

(I.e. notes about implementation and interface are seperate)

In which case, it seems reasonable to me that the docstring
on an function named 'xmlrpc_foo' contains information on
how to call it from the context of XML-RPC -- it's unlikely
that it would be invoked any other way.

This is unlike Zope 2 (which I also hate), where the same
method is often intended for both remote use and local use.
   And in Zope 2, docstrings affect the ability to *call* a
function, but this only impacts introspection.  Also, given
that assigning approriate function attributes would override
the details extracted from the docstring (if I understand
the patch correctly), *and* you could plug in a completely
different introspector if you wanted to anyway (whereas the
docstring gunk in Zope is hard-wired at low levels), I don't
see this as a problem at all -- instead, it's a convenient
and sensible default behaviour that makes automatic use of
the docstrings that developers will probably write anyway.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Oh, and make the example code use reactor (otherwise we'll
just need to change it again in 1.0.8 when
twisted.application hits the codebase.)

And, "self.xmlrpc_parent = parent", won't that allow xml-rpc
calls to system.parent which will then break in a peculiar way?

But feel free to commit. Thanks for all the work you put
into this!

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
Doh! You're right, I'll use self._xmlrpc_parent.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
Just for the record, here is the final patch.

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_37"></a>|@glyph commented|
|-|-|

In changeset c5d7c6bad4ca26457c120ca3026d81577ed38514
```
#!CommitTicketReference repository="" revision="c5d7c6bad4ca26457c120ca3026d81577ed38514"
Merge pull request #80 from rodrigc/has_key6

Author: rodrigc

Reviewer: glyph

Fixes: #8364

Replace dict.has_key() with the 'in' keyword for twisted/web, to prepare for migration for Python 3.
```
